flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java

```diff
@@ -141,25 +141,29 @@ public void stop() {
 					final ApplicationStatus applicationStatus;
 					if (t != null) {
 
-						final Optional<JobCancellationException> cancellationException =
-								ExceptionUtils.findThrowable(t, JobCancellationException.class);
-
-						if (cancellationException.isPresent()) {
+						if (ExceptionUtils.findThrowable(t, JobCancellationException.class).isPresent()) {
 							// this means the Flink Job was cancelled
 							applicationStatus = ApplicationStatus.CANCELED;
-						} else if (t instanceof CancellationException) {
-							// this means that the future was cancelled
-							applicationStatus = ApplicationStatus.UNKNOWN;
-						} else {
+							LOG.warn("Application {}: ", applicationStatus, t);
+
+							return dispatcher.shutDownCluster(applicationStatus);
+						}
+
+						if (ExceptionUtils.findThrowable(t, JobExecutionException.class).isPresent()) {
```

Comment:
Are you sure that a `JobExecutionException` means that the Flink job has reached a globally terminal state? Looking at `JobResult.toJobExecutionResult` it seems that we also throw this exception if the state is `ApplicationStatus.UNKNOWN`.